[WIP] Relayer ADR and diagrams (#28)
informalsystems/ibc-rs: docs/architecture/adr-002-ibc-relayer.md

# ADR 002: IBC Relayer in Rust

## Changelog
* {date}: {changelog}

## Definitions
On-chain IBC Client (or IBC Client) - client code running on chain, typically only the light client verification related functionality.

Relayer Light Client - full light client functionality, including connecting to at least one provider (full node), storing and verifying headers, etc.

Source chain - the chain from which the relayer reads data to fill an IBC datagram.

> Destination chain - the chain where the realyer submits transactions that include the IBC message.

Reviewer: ```suggestion
Destination chain - the chain where the relayer submits transactions that include the IBC message.
```

A and B chains - for connection protocol, A is the "initiating" chain where `MsgConnectionOpenInit` is initially processed and eventually `MsgConnectionOpenAck`. B is the chain where `MsgConnectionOpenTry` and `MsgConnectionOpenConfirm` are processed. 
Similar for channel handshake protocol.

## Context

> A relayer is an off-chain process responsible for relaying IBC packet data & metadata between two or more chains by scanning their states and submitting transactions. This is because in the IBC architecture, modules are not directly sending messages to each other over networking infrastructure, but instead they create and store the messages to be retrieved and forwarded by a relayer process. 

Author: IBC packets here means anything IBC...but I can see how this can be confusing. Will add definitions.

#### Data Availability
The relayer monitors the chain state to determine when packet forwarding is required. The relayer must be able to retrieve the data within some time bound. This is referred to as **data availability**.

#### Data Legibility
IBC protocol defines the minimal data set that must be made available to relayers for correct operation of the protocol. The relayer expects the data to be legible, i.e. **data should be serialized** according to the IBC specification format; this includes consensus state, client, connection, channel, and packet information, and any auxiliary state structure necessary to construct proofs of inclusion or exclusion of particular key/value pairs in state. 
 - [IBC Specification] some protobuf specifications can be found under individual ICS-es, for exmple [ICS-03 connection protobufs](https://github.com/cosmos/ics/blob/master/spec/ics-002-client-semantics/data-structures.proto)

> #### IBC Querier Module

Author: I changed the title ;)

> #### IBC Querier Module

Author: > do we meant to query the logs or what exactly?

There are requirements to provide `query` functions for most of the  IBC Modules (client, connection, channel, packet, etc). This section says that there should be an implementation available. Note that this should be in the Rust Modules implementation for the "front-end" while the backend is Cosmos-SDK module implementation.

IBC host state machines MUST provide query functions for all data available to relayers. For Cosmos/Tendermint chains this means:
- the IBC modules on chain correctly implement and respond to queries
  - [IBC-Modules-Rust] an implementation for some queries currently exist in Cosmos-SDK and same and more need to be implemented in Rust. The full requirements are detailed in section Relayer Queries. 

> - the relayer needs the ability to send rpc/http ABCI queries to and receive replies from Tendermint/Cosmos-SDK

Author: No, this is just to be able to send a tendermint (query) request and receive a response.

  - [IBC-Modules-Rust] identifier validation is required (ICS-024)
  - [IBC-Modules-Rust] requires Rust types for all query responses
  - [[Merkle-Proofs-Rust](https://github.com/confio/ics23/tree/master/rust)] (candidate implementation) - some query responses include proofs and included in IBC transactions by the relayer (some may be validated, TBD)

#### IBC Messages
The relayer creates transactions that include IBC messages to manage clients, connections and channels, and send application packets to destination chains. These messages must be defined in the IBC Rust implementation [IBC-Modules-Rust].

#### IBC Logging System
IBC packet data & timeouts are not stored directly in the chain state (as this storage is presumed to be expensive) but are instead committed to with a succinct cryptographic commitment (only the commitment is stored).
As a consequence, IBC requires that a **host state machine MUST provide an event logging system** that logs data in the course of transaction execution. **Logs must be queryable** by relayers to read IBC packet data & timeouts. 

The logging system must provide the following functions:

>  - [IBC-Modules-Go] emitLogEntry for emitting log entries called by the state machine during transaction execution:

Reviewer: Why reference to Go?

Author: in the initial version we will test with Cosmos-SDK/TM in Go chains.

   - type emitLogEntry = (topic: string , data: []byte) => void
   - example: emitLogEntry("sendPacket", {sequence: packet.sequence , data: packet.data, timeout: packet.timeout})
 - [IBC-Modules-Go] queryByTopic for querying past logs matching a given topic:
   - type queryByTopic = (height: uint64 , topic: string) => Array < []byte >

#### Keyring 
The relay process must have access to its accounts with tokens on all destination chains, with sufficient balance to pay for transaction fees. Account key information must be stored and managed securely. A keyring implementation is required for CRUD key operations. 
[Keyring-Rust] Investigation in existing Rust implementations is needed. (ex: [hwchen-keyring](https://github.com/hwchen/keyring-rs))

### Chain Transactions and Signing
The relayer must create chain specific signed transactions. 
[Cosmos-Tx-Rust] For the first release Cosmos-SDK transaction signing is required. One possible implementation is [iqlusion's sdtx crate](https://github.com/iqlusioninc/crates/tree/develop/stdtx)

#### Implementation of IBC "routing module"

> The default IBC handler uses a receiver call pattern, where modules must individually call the IBC handler in order to bind to

Reviewer: This part is not very clear.

Reviewer 2: Had some issues with this as well, until I read [this](https://github.com/cosmos/ics/blob/a4173c91560567bdb7cc9abee8e61256fc3725e9/ibc/5_IBC_DESIGN_PATTERNS.md#call-receiver).

> ports, start handshakes, accept handshakes, send and receive packets, etc. While this provides flexibility for modules, it imposes extra work on the part of the relayer processes that now needs to track the state of multiple modules. The IBC specification describes an IBC “routing module” to route packets, and simplify the task of relayers. This routing module accepts external datagrams and calls into the IBC handler to deal with handshakes and packet relay. The routing module keeps a lookup table of modules, which it can use to look up and call a module when a packet is received, so that external relayers need only ever relay packets to the routing module.

Reviewer: "so that external relayers need only ever relay packets to the routing module". I don't understand this sentence: I guess that relayer has option of sending transaction to a chain, not to a particular module. Once transaction with IBC packet is received, it goes to routing module that is then forwarding it to the right module. Is this true? And who is responsible for determining if transaction contains IBC packet data?

Author: > Is this true? 

Yes
> And who is responsible for determining if transaction contains IBC packet data?

This depends on the chain implementation. In Cosmos-SDK, for application packets, it's going to be the implementation of ICS-026 (this routing module) that keeps track of the different application modules, their channels and ports, and do the correct handling. For handshake messages this is going to be the IBC modules.

[IBC-Routing-Module-Go] Initial version of the relayer assumes that chains implement the "routing module"

#### Batching

> The relayer may batch transactions if supported by destination chain and allowed by configuration. In this case the relayer  can bundle multiple datagrams into a single transaction and amortise any overhead costs (e.g. signature checks for fee payment).

Reviewer: Is relayer batching IBC packets in a transaction or multiple transactions in a single transaction?

> The relayer may batch transactions if supported by destination chain and allowed by configuration. In this case the relayer  can bundle multiple datagrams into a single transaction and amortise any overhead costs (e.g. signature checks for fee payment).

Reviewer: ```suggestion
The relayer may batch IBC datagrams in a single transaction if supported by destination chain and allowed by configuration. In this case the relayer can  amortise any overhead costs (e.g. signature checks for fee payment).
```

Initial version of the relayer assumes batching is supported by all chains. An option may be later included in the configuration file.

## Relayer Requirements

A correct relayer MUST:

- **[R-config-start]** Read, parse, validate a configuration file upon start and configure itself for the specifed chains and paths
- **[R-transport]** Have access to the networking protocols (e.g. TCP/IP, UDP/IP, or QUIC/IP) and physical transport, required to read the state of one blockchain/ machine and submit data to another

> - **[R-provider]** Maintain transport connections to at least one full node per chain

Reviewer: This can be turned into maintain light client connections to each chain (and that implies this requirement).

Author: I will clarify. client connection is used to get headers, the other (need a better name) is for queries and transactions (these don't go via light client). Same full node can and will be used by both.

- **[R-query]** Query IBC data on source and destination chains
- **[R-light-client]** Run light clients for source chains and 

> - **[R-IBC-client]** create and update IBC clients on destination chains 

Reviewer: We need IBC clients on both source and destination chains (because of Connection Handshake protocol).

Author: Yes but if this relayer is not configured to relay to source, this should be done by another relayer. I was trying to describe just a unidirectional relay but I see this is more confusing. Will figure out how to change.

- **[R-accounts]** Own accounts on destination chains with sufficient balance to pay for transaction fees

> - **[R-transact]** Create, sign and forward IBC datagram transactions

Reviewer: Might be good idea defining IBC datagram transaction (a transaction that contains IBC packets (we use datagram term sometimes).

Author: attempted a fix here

- **[R-relay]** Perform correct relaying of all required messages, according to the IBC sub-protocol constraints
- **[R-restart]** Resume correct functionality after restarts
- **[R-upgrade]** Resume correct functionality after upgrades
- **[R-proofs]** Perform proof verification (as it will be done on the destination chain) and not forward messages where proof verification fails

The relayer MAY:
- **[R-config-cli]** Provide ways to change configuration at runtime

> - **[R-bisection]** Perform bisection to optimize transaction costs and computation on destination chains

Reviewer: I guess what is meant here is try to identify bisection chains (list of headers) that have smaller transaction costs; as bisection is executed by default by light client.

Author: the full light client that executes on the relayer performs bisection by default. In one example the relayer who uploads headers periodically,  could have headers 1, 4, 7, 10, etc. from A. But the relayer doesn't need to update client on B if there is no IBC datagram to be relayed. If some IBC data is created at height 11 on A then the relayer can determine which headers need to be sent to B out of the 1, 4, 7, 10, 11 by simulation bisection on (th=1, sh=11).

- **[R-relay-prio]** Filter or order transactions based on some criteria (e.g. in accordance with the fee payment model)

## Implementation
The initial implementation will heavily borrow from the Go relayer implementation that uses a "naive" algorithm for relaying messages. The structure of the configuration file is similar with the one in Go (see [Go-Relayer](https://github.com/cosmos/relayer))

### Configuration
(WIP)
Upon start the relayer reads a configuration file that includes global and per chain parameters. The file format is .toml
The relayer performs initialization based on the content of this file. Below is an example of a configuration file.

```$xslt
# This is an IBC relayer sample configuration
title = "IBC Relayer Config Example"

[global]
timeout = "10s"
strategy = "naive"

[[chains]]
  id = "chain_A"
  rpc_addr = "http://localhost:26657"
  account_prefix = "cosmos"
  key_name = "testkey"
  store_prefix = "ibc"
  client_ids = ["clA1", "clA2"]
  gas = 200000
  gas_adjustement = 1.3
  gas_price = "0.025stake"

>   trusting_period = "336h"

Reviewer: It might be more clear (and clean) if parameters are grouped (for example light client related goes under light client section, chain related under chain section, etc).

Author: Yes, the config file will change quite a bit probably. We will have an issue/PR just for this.

[[chains]]
  id = "chain-B"
  rpc_addr = "http://localhost:26557"
  account_prefix = "cosmos"
  key_name = "testkey"
  store_prefix = "ibc"
  client_ids = ["clB1"]
  gas = 200000
  gas_adjustement = 1.3
  gas_price = "0.025stake"
  trusting_period = "336h"

[[connections]]

[connections.src]

>   client_id = "clA1"

Reviewer: How we know on what chain this endpoint is created?

Author: on the chain that "defines" that client-id (i.e. in the `client_ids =..` section of that chain)

  connection_id = "conn1-idA-clA1"
[connections.dest]
  client_id = "clB1"
  connection_id = "conn1-idB-clB1"

[[connections.paths]]
  src_port = "app1-port-A"
  dest_port = "app1-port-B"
  direction = "unidirectional"

[[connections.paths]]
  src_port = "app2-port-A"
  dest_port = "app2-port-B"
  direction = "bidirectional"
```

#### Global
```
[global]
timeout = "10s"
strategy = "naive"
```

Relaying is done periodically and the frequency is dictated by the `timeout` parameter. The `strategy` parameter configures the relayer to run a particular relaying algorithm.

#### Chains 
Chain level information including account and key name, gas information, trusting period, etc. All source and destination chains must be listed here.
```
[[chains]]
  id = "chain-B"
  rpc_addr = "http://localhost:26557"
  account_prefix = "cosmos"
  key_name = "testkey"
  store_prefix = "ibc"
  client_ids = ["clB1"]
  gas = 200000
  gas_adjustement = 1.3
  gas_price = "0.025stake"
  trusting_period = "336h"
```

#### Relay Paths 

> The realyer may be configured to relay between some application ports, over a number of connections and channels, in unidirectional or bidirectional mode.

Author: will do!

- `query_all_client_states(chain)` - returns the IBC light clients instantiated on the chain
- `query_client_consensus_state(chain, clientID, height)` - returns the consensus state proof for a light client at a given height if height > 0, else it returns the latest height
- `query_connections(chain)` - returns all connections created on the chain
- `query_client_connections(chain, clientID)` - returns all connections associated with a light client
- ...more to be added

### Relayer Algorithm
A relayer algorithm is described in [relayer algorithm described in IBC Specifigication](https://github.com/cosmos/ics/blame/master/spec/ics-018-relayer-algorithms/README.md#L47) and [Go relayer implementation ](https://github.com/cosmos/relayer/blob/f3a302df9e6e0c28883f5480199d3190821bcc06/relayer/strategies.go#L49.).

This section describes the details of the relayer algorithm in the Rust implementation. 
The purpose of the relayer algorithm is to build and forward messages over the `src->dest` relaying paths it has been configured for. It is executed periodically with the frequency specified in the configuration file, with each run being a **relay cycle**.

> For every destination the relayer:

Reviewer: Not clear what is meant by destination here. Is it for every chain (defined in config file)?

- queries clients, connection, channels and packet related state on source and destination chains,
- creates new datagrams if needed, 
- batches multiple datagrams in single transactions, 

> - signs and submits these datagrams to the destination.

Reviewer: ```suggestion
- signs and submits these transactions to the destination.
```

  clientMsgsForB = []
  // get latest header and height on A
  heightA = A.latestHeight()
  headerA = A.latestHeader()

  // get all clients instantiated on B
  existingOnB = query_all_client_states(B)

  for clIdA = range config.A.client_ids {
    if clIdA not in existingOnB {
      // create client 
      msg = ClientCreate({clIdA, headerA, config.A.trusting_period, config.A.unbonding_period, config.A.Signer()})

    } else if query_client(B, "A_store/clients/clIdA").height < heightA {
      // update client
      msg = ClientUpdate({clIdA, headerA, config.A.Signer()})

    } else {
      // do nothing
      continue
    }

    clientMsgsForB.push(msg)
  }
}
```

The relayer runs its own light client for A, retrieves and verifies headers, and updates the A-client on chain B with new headers as required. 

Since the relayer must pay for all transactions, including `clientCreate` and `clientUpdate` ones, there are incentives for optimizations.
For example, light client implementation of Tendermint supports bisection and the relayer may choose to send skipping headers to A-client on B, periodically or when required by new IBC datagrams.

#### Connection Messages

> The relayer queries the source and destination chains of the relaying paths in order to determine if connection handshake datagrams should be sent to destination chain.

Author: It is possible that the relayer is not configured to relay to source chain.

##### Connection Query

The following structures pertain to connection queries and should be detailed in [IBC-Modules-Rust-ADR]. 
The structures are shown here for reference. (??)

[todo - make it Rust]
```typescript
type Counterparty struct {
	ClientID     string            
	ConnectionID string           
	Prefix       commitment.PrefixI 
}
type ConnectionEnd struct {
	State    exported.State 
	ClientID string 
	Counterparty Counterparty 
	Versions []string 
}
// ConnectionResponse defines the client query response for a connection.
// It includes the proof and the height at which the proof was retrieved.
type ConnectionResponse struct {
	Connection  ConnectionEnd    
	Proof       commitment.Proof 
	ProofPath   commitment.Path  // specified in ICS024
	ProofHeight uint64           
}
```

> The correct connection states on the `src` and `dest` chains must be verified by the relayer. In addition the relayer may perform some proof verification. 

Author: There is none. It's just an option for the relayer to perform the proofs the destination chain does in order to ensure it doesn't pay for things that will fail.

A number of IBC datagrams contain proofs obtained from chain A at a some height `h`. A proof needs to be verified on B against the commitment root for `h` which, for Tendermint clients, is included in the client consensus state at `h+1`. This is because for Tendermint chains the application Hash after applying all transactions in block `n` is included in block at height `n+1`. 

The relayer therefore needs to ensure that the consensus state at `proof_height+1` exists on chain B.

One proposal is shown below and described in the rest of this section.
![IBC_client_heights](assets/IBC_client_heights.jpeg)

The relayer creates a light client on B with `hi` and then updates it as required by processing different IBC events. Let `ha'` be the last consensus state for client on B.
When some IBC event for X (connection, channel or packet) is received, it includes the height, let it be `hx-1` at which the event occured on A.
According to the proposal here, the relayer should:
- get the latest consensus state height of client on B, `ha`
- let `h = max(hx, ha)`
- query for item X at height `h-1` and get a proof `p` at this height

> - wait for the block at height `hx` to be received, i.e. `Ev{bloack, hx}`

Reviewer 3: Found while skimming through the doc:
```suggestion
- wait for the block at height `hx` to be received, i.e. `Ev{block, hx}`
```

* {date}: {changelog}

## Definitions
These definitions are specific for this document and they may not be consistent with the IBC Specification.

> IBC datagram - the transaction payload sent by the relayer over the physical network that includes client, connection, channel and IBC packet data. 

Reviewer 2: A small clarification would make this definition amazing, namely: Where does the relayer send an IBC datagram to? i.e., what is the physical network endpoint delivering and handling an IBC datagrams from a relayer? Is this is a full node, for instance in Tendermint accepted through the `broadcast_tx_commit` API?

IBC packet - a particular type of IBC datagram that includes the application packet and its commitment proof.

> IBC datagram transaction - a transaction that includes an IBC datagram.

Reviewer 2: (This suggestion is rather a test to see if I understood what you mean.)
```suggestion
IBC datagram transaction - an on-chain transaction that includes an IBC datagram.
```

Author: Why on-chain? We talk about this before it is submitted to the destination chain
